File: algorithms/2_medium/P/00131_palindrome_partitioning/backtrack.cpp
typedef ptrdiff_t isize;

class Solution {
    static inline bool is_palindrome(const span<const char> s) {
        isize hi = std::ssize(s) - 1;
        isize lo = 0;

        while (lo < hi) {
            if (s[lo] != s[hi]) {
                return false;
            }

            lo += 1;
            hi -= 1;
        }

        return true;
    }

    static void backtrack(const string_view s, const isize lo, vector<vector<string>>& res, vector<string>& temp) {
        const isize len = std::ssize(s);

        if (lo >= len) {
            res.push_back(temp);
            return;
        }

        for (isize hi = lo; hi < len; hi += 1) {
            if (is_palindrome(span<const char>(&s[lo], hi - lo + 1))) {
                temp.push_back(string(s.cbegin() + lo, s.cbegin() + hi + 1));
                backtrack(s, hi + 1, res, temp);
                temp.pop_back();
            }
        }
    }
public:
    static vector<vector<string>> partition(const string& s) {
        auto res = vector<vector<string>>();
        auto temp = vector<string>();

        backtrack(s, 0, res, temp);

        return res;
    }
};

File: algorithms/2_medium/P/00131_palindrome_partitioning/backtrack.rs
impl Solution {
    #[inline(always)]
    fn is_palindrome(bytes: &[u8]) -> bool {
        assert!(!bytes.is_empty());
        let mut hi = bytes.len() - 1;
        let mut lo = 0_usize;

        while lo < hi {
            if bytes[lo] != bytes[hi] {
                return false;
            }

            lo += 1;
            hi -= 1;
        }

        true
    }

    fn backtrack(bytes: &[u8], lo: usize, res: &mut Vec<Vec<String>>, temp: &mut Vec<String>) {
        let len = bytes.len();

        if lo >= len {
            res.push(temp.clone());
            return;
        }

        for hi in lo..len {
            if Self::is_palindrome(&bytes[lo..=hi]) {
                temp.push(String::from_utf8(bytes[lo..=hi].to_vec()).unwrap());
                Self::backtrack(bytes, hi + 1, res, temp);
                temp.pop();
            }
        }
    }

    pub fn partition(s: String) -> Vec<Vec<String>> {
        let mut res: Vec<Vec<String>> = Vec::default();
        let mut temp: Vec<String> = Vec::default();

        Self::backtrack(s.as_bytes(), 0, &mut res, &mut temp);

        res
    }
}
